### PR TITLE
port roi cluster fusion

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

- port roi_cluster_fusion package
- use **DetectedObjectsWithFeature** (https://github.com/tier4/AutowareArchitectureProposal_msgs/pull/3) instead of DynamicObjectWithFeatureArray
- subscribe topic type: DynamicObjectWithFeatureArray -> **DetectedObjectsWithFeature**
- publish topic type: DynamicObjectWithFeatureArray -> **DetectedObjects**

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [pull request guidelines][pull-request-guidelines]
- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp [hunk 1]
```diff
@@ -243,30 +243,34 @@ void RoiClusterFusionNodelet::cameraInfoCallback(
...
   }
 
   // build output msg
-  autoware_perception_msgs::msg::DynamicObjectWithFeatureArray output_msg;
-  output_msg = *input_cluster_msg;
+  autoware_auto_perception_msgs::msg::DetectedObjects output_msg;
+  output_msg.header = input_cluster_msg->header;
+  for (const auto & feature_object : input_cluster_msg->feature_objects) {
+    output_msg.objects.push_back(feature_object.object);
+  }
```
Comment (satoshi-ota) [inline perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp:266]: @yukkysaito @yukke42  please check the following modifications :pray: 

- change input topic type DynamicObjectWithFeatureArray -> DetectedObjectsWithFeature
- change output topic type DynamicObjectWithFeatureArray -> DetectedObjects
- remove autoware_perception_msgs/Feature from DetectedObjectsWithFeature in this process

Comment (satoshi-ota) [reply perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp]: fix output topic type in https://github.com/tier4/AutowareArchitectureProposal.iv/pull/536/commits/fb8c160d15f031962b06b6cac48d56098b7083d5

- DetectedObjects -> DynamicObjectWithFeatureArray

File: perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp [hunk 2]
```diff
@@ -416,10 +417,12 @@ void RoiClusterFusionNodelet::fusionCallback(
...
       if (
         iou_threshold_ < max_iou &&
-        output_msg.feature_objects.at(index).object.semantic.confidence <=
-          input_roi_msg->feature_objects.at(i).object.semantic.confidence) {
-        output_msg.feature_objects.at(index).object.semantic =
-          input_roi_msg->feature_objects.at(i).object.semantic;
+        output_msg.feature_objects.at(index).object.existence_probability <=
+          input_roi_msg->feature_objects.at(i).object.existence_probability &&
+        input_roi_msg->feature_objects.at(i).object.classification.front().label !=
+          autoware_auto_perception_msgs::msg::ObjectClassification::UNKNOWN) {
```
Comment (satoshi-ota) [inline perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp:423]: @yukkysaito @yukke42 please check logic modification :pray: 

- overwrite classification when it satisfies following conditions:
- roi's existence probability is larger than cluster's one
- roi's classification label is **NOT** UNKNOWN

File: perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp (hunk 1 above)
Comment (yukkysaito) [reply perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp]: LGTM

File: perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp (hunk 2 above)
Comment (yukkysaito) [reply perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp:423]: LGTM

Comment (yukkysaito) [reply perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp:423]: （indexとiは気持ち悪いのでどっかで整理したいですね）

**Review by yukkysaito — APPROVED** [review]

File: perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp (hunk 2 above)
Comment (satoshi-ota) [reply perception/object_recognition/detection/roi_cluster_fusion/src/roi_cluster_fusion_nodelet.cpp:423]: memo at porting spread sheet :+1:

**Pull request merged**